iree/compiler/JitEval/Runtime.cpp

```diff
@@ -0,0 +1,232 @@
+  Builder builder(loc.getContext());
+  if (iree_hal_element_numerical_type_is_integer(halElementType)) {
+    return builder.getIntegerType(iree_hal_element_bit_count(halElementType));
+  } else if (halElementType == IREE_HAL_ELEMENT_TYPE_FLOAT_32) {
+    return builder.getF32Type();
+  } else if (halElementType == IREE_HAL_ELEMENT_TYPE_FLOAT_64) {
+    return builder.getF64Type();
+  } else if (halElementType == IREE_HAL_ELEMENT_TYPE_FLOAT_16) {
+    return builder.getF16Type();
+  } else if (halElementType == IREE_HAL_ELEMENT_TYPE_BFLOAT_16) {
+    return builder.getBF16Type();
+  }
+
+  emitError(loc) << "unrecognized evaluated buffer view element type: "
+                 << halElementType;
+  return {};
+}
+
+}  // namespace
+
+CompiledBinary::CompiledBinary() {
+  IREE_CHECK_OK(iree_vm_instance_create(iree_allocator_system(), &instance));
+}
+
+CompiledBinary::~CompiledBinary() {
+  iree_vm_context_release(context);
+  iree_vm_module_release(main_module);
+  iree_vm_instance_release(instance);
+}
+
+LogicalResult CompiledBinary::invokeNullary(Location loc, StringRef name,
+                                            ResultsCallback callback) {
+  iree_vm_function_t function;
+  IREE_CHECK_OK(main_module->lookup_function(
+      main_module->self, IREE_VM_FUNCTION_LINKAGE_EXPORT,
+      iree_string_view_t{name.data(), name.size()}, &function));
+
+  iree::vm::ref<iree_vm_list_t> inputs;
+  IREE_CHECK_OK(iree_vm_list_create(/*element_type=*/nullptr, 0,
+                                    iree_allocator_system(), &inputs));
+  iree::vm::ref<iree_vm_list_t> outputs;
+  IREE_CHECK_OK(iree_vm_list_create(/*element_type=*/nullptr, 1,
+                                    iree_allocator_system(), &outputs));
+
+  IREE_CHECK_OK(iree_vm_invoke(context, function, IREE_VM_INVOCATION_FLAG_NONE,
+                               /*policy=*/nullptr, inputs.get(), outputs.get(),
+                               iree_allocator_system()));
+
+  if (failed(callback(outputs.get()))) {
+    return failure();
+  }
+  return success();
+}
+
+Attribute CompiledBinary::invokeNullaryAsElements(Location loc,
+                                                  StringRef name) {
+  Attribute result;
+  if (failed(invokeNullary(
+          loc, name, [&](iree_vm_list_t* outputs) -> LogicalResult {
+            if (iree_vm_list_size(outputs) != 1) {
+              return emitError(loc) << "expected 1 result for func " << name
+                                    << " got " << iree_vm_list_size(outputs);
+            }
+            iree_vm_variant_t variant = iree_vm_variant_empty();
+            IREE_CHECK_OK(iree_vm_list_get_variant(outputs, 0, &variant));
+            result = convertVariantToAttribute(loc, variant);
+            return success(result != nullptr);
+          }))) {
+    return nullptr;
+  }
+
+  return result;
+}
+
+Attribute CompiledBinary::convertVariantToAttribute(
+    Location loc, iree_vm_variant_t& variant) {
```

Comment:
this is really cool - being able to evaluate anything and not just those that result in tensors!